File: D_star Lite/D_star Lite_5cm/UpdateVertex.m
function UpdateVertex(u)%更新节点信息
    global rhs;
    global g;
    global U;
    try
        [IsInFlag,index]=isInOpenList(u,U);
    catch
        IsInFlag=false;
    end
    if g(u(1),u(2))~=rhs(u(1),u(2)) && IsInFlag
        % U.Update
        [key(1),key(2)]=CalculateKey(u);
        U(index,3:4)=key;

    elseif g(u(1),u(2))~=rhs(u(1),u(2)) && ~IsInFlag
        [calculated_key(1),calculated_key(2)]=CalculateKey(u);
        U=[U
           u,calculated_key];
    elseif g(u(1),u(2))==rhs(u(1),u(2)) && IsInFlag
        U(index,:)=[];
    end
end

File: D star Lite C++/D_start_Lite/src/UpdateVertex.cpp
/* Include files */
#include "main.h"
#include "Initialize.h"
#include "UpdateVertex.h"
#include "isInList.h"
#include "CalculateKey.h"

/* Parameters Declarations */
static int isInListFlag;
static int isInListIndex;

/* Matrix Declarations */
static RowVectorXi result_array;
static RowVector2d key;
static RowVectorXd temp_row(4);

/* Function Declarations */
void UpdateVertex(RowVectorXd u)
{
    try
    {
        result_array = isInList(u, U);
        isInListFlag = result_array(0);
        isInListIndex = result_array(1);
    }
    catch(...)
    {
        isInListFlag = 0;
    }

    if ((g(u(0), u(1)) != rhs(u(0), u(1))) && isInListFlag)
    {
        // U.Update
        key = CalculateKey(u);
        //更新key值
        U(isInListIndex, 2) = key(0);
        U(isInListIndex, 3) = key(1);
    }
    else if ((g(u(0), u(1)) != rhs(u(0), u(1))) && !isInListFlag)
    {
        //计算key值后加入U
        key = CalculateKey(u);
        temp_row << u(0), u(1), key;
        appendRow(U, temp_row);
    }
    else if ((g(u(0), u(1)) == rhs(u(0), u(1))) && isInListFlag)
    {       
        //从U中删除节点
        removeRow(U, isInListIndex);
    }



}